### PR TITLE
Topic recognition (46980801): GNN

### PR DESCRIPTION
**Pull request by NJChay**
This pull request adds an implementation of a Graph Neural Network to the topic recognition folder. The project contains a dataset.py file for loading and preprocessing data, a modules.py file for the definition of the model architecture, a train.py file where the model is trained on the data with selected parameters and finally a predict.py file where test inputs are run through the model and visualisations of output are created and saved. The project also contains a graphs folder where all the training plots and output visualisations have been stored and a README that describes the motivation, implementation and usage of the project files.

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**
GNN FB Page-Page Network dataset ---- Normal Difficulty
| Category	| | Marks | Comments |	
| :---  | :---: | :---: | :---  |		
| Algorithm solves the problem 		| 5 | 4 | No test/train loss plots. No early stop.  |
| Implementation functions as intended	| 3 | 3 | 4 layer GCN, torch_geometric/GCNConv |
| Good design 					| 1 | 1 | Modular, Reusable |
| Commenting 					| 1 | 1 | Meaningful docstrings and comments |
| Algorithm above Normal Difficulty 	| 5 | 5 | |
| Algorithm is Hard difficulty 			| 5 | 0 | Normal Difficulty |
| Section IV : Max mark 15 from 20		|   | **14**   | |

**TSNE / UMAP:** 
Good.
- Discussion: Good.

**Suggestions:**
Hard coded hyper parameters should be avoided.
Code can have some early stop, ex: train loop should save best model by keeping track of current best accuracy.
Could add a hyper parameter search or discuss how the values were obtained.

**Comment by NJChay** [comment]
Updated train function with early stopping and appropriate loss plots as per feedback

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. </td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td>-1</td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. GNN Task.</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is adequate. Add some info about the specific implementation here.</td><td>-1</td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-9</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by NJChay** [comment]
It seems I have lost an additional two marks from the initial inspection? Including losing an extra mark for having an okay solution to the problem when I achieved >90% accuracy on the test set and a perfect TSNE plot? I would appreciate some further clarification of this as it is important as I am now one mark under the 80% threshold for a seven due to these extra lost marks.

**Comment by hanemma7moud** [comment]
Nicholas Chay
